ecosystem/op-deployer.md

```diff
@@ -0,0 +1,163 @@
+so read that document first if you haven't already. This design doc will focus on the implementation details of the
+architecture it describes.
+
+## Deployment Pipeline
+
+The core of the modular deployment tooling is the concept of a _deployment pipeline_. The pipeline consists of stages
+that each perform a single piece of the deployment. The stages each take an input file, and enrich it with the data
+pertaining to that particular stage of the deployment. This allows the output of upstream stages to be cached.
+
+The diagram below outlines what these stages are:
+
+![](./op-deployer/pipeline.png)
+
+## Deployment Intents
+
+We'll create a new tool called `op-deployer` to orchestrate deployments. Users won't need to interact with the pipeline
+directly. Instead, they'll define a _deployment intent_ which will describe the properties of the chains they want to
+deploy. `op-deployer` will then diff the intent against the current state of the deployment, and run the pipeline to
+ensure that the real deployment matches the intent.
+
+An example deployment intent is below. Deployment intents are encoded as TOML to match what the OP Stack Manager
+expects:
+
+```toml
+[intent]
+l1ChainID = 11_155_111
+useFaultProofs = true
+fundDevAccounts = true
+# Define the version of the smart contracts to deploy
+contractsVersion = "op-contracts/1.2.0"
+# Override specific variables in the deploy config globally
+overrides = { }
+
+# Define the chains to deploy, by chain ID
+[intent.chains.901]
+# Override variables in the deploy config for this specific chain
+overrides = { }
+
+[intent.chains.902]
+# ... etc.
+```
+
+Under the hood, `op-deployer` will run through the deployment pipeline. The output of each intermediate stage will
+be stored by default within the intent's TOML file, however the implementation of the "state store" will be
+pluggable in order to support use cases like `opc` which store the deployment state in a database. Example output from a
+stage is below:
+
+```toml
+[intent]
+# Intent data is same as above and elided
+
+[state.chains.901]
+# Ownership addresses derived from a mnemonic
+proxyAdminOwner = "0xe59a881b2626f948f56f509f180c32428585629a"
+finalSystemOwner = "0xb9cdf788704088a4c0191d045c151fcbe2db14a4"
+baseFeeVaultRecipient = "0xbc4a9110dad00b4d9fb61743598848ddda6eeb03"
+l1FeeVaultRecipient = "0x6b0c2542fa2cadced5c7f64ef6fb9ebbce7630ff"
+sequencerFeeVaultRecipient = "0xb4e5b724bbc54c95d292613d956871281120ead6"
+
+# Address of OPSM so that other addresses can be retrieved from on-chain data
+opStackManagerAddress = "0x79c6c6b1844e3db7c30107f189cfb095bd2c4b5d"
+
+# Genesis data
+genesis = "base64://abcd..."
+genesisHash = "0x1234..."
+genesisTime = 1234567890
+
+[state.chains.902]
+# ... etc.
+```
+
+Only the minimum necessary state is stored. For example, the rollup config can be derived from deploy config
+variables and therefore can be generated on-the-fly when it is needed. `op-deployer` will expose utility subcommands
```

Comment:
Nit: It may be nice to have a command that outputs all the configs and deploy artifacts exhaustively so the complete state can be readily examined / debugged as needed.